src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -1212,23 +1275,77 @@ private bool AddIfNotPresent(T value)
                 Initialize(0);
             }
 
-            int hashCode = InternalGetHashCode(value);
-            int bucket = hashCode % _buckets!.Length;
+            int hashCode;
+            int bucket;
             int collisionCount = 0;
             Slot[] slots = _slots;
-            for (int i = _buckets[bucket] - 1; i >= 0; i = slots[i].next)
+
+            IEqualityComparer<T>? comparer = _comparer;
```

Comment:
Same suggestions as above for merging code paths and calculating hashCode in the common path.